### ISSUE TITLE
Allow adding code coverage data collected from other processes before dumping report

### ISSUE THREAD

**Issue by TheCelavi** [sebastianbergmann/phpunit#5130]
We are working on general purpose websocket server using Swoole. Now, that is pretty much complicated topic for anyone not familiar with Swoole, but in general, Swoole spawns workers using process forks (not main thread, not possible to write e2e tests in main thread, like, per example, Symfony does).

It is quite possible to collect coverage report for each worker using `pcov` and to generate and serialize instance of `\SebastianBergmann\CodeCoverage\RawCodeCoverageData` for each individual worker. 

By simple code modification of class `\PHPUnit\TextUI\TestRunner` I have managed to combine code coverage from main PHPUnit thread which executed unit tests WITH code coverage of Swoole workers which were subject of E2E tests using PHPUnit.

Now, I can not stress enough how much value we got with this one, combined, report.

However, in order to achieve that, we had to "violate" TestRunner class. 

It would be AWSOME if PHPUnit would add an extension point, a new interface, something like this:

```php
interface BeforeCodeCoverageReportGeneratedHook {

         execute beforeCodeCoverageReportGenerated(\SebastianBergmann\CodeCoverage\CodeCoverage $coverage): void;

}
```

So we could in that extension execute `CodeCoverage::append($collected, 'e2e');`

This would allow us clean method to have full code coverage report containing data from unit tests as well as E2E tests.

Of course, if you are willing to accept this feature request, I would be very happy to send PR. 

Thanks.

**Comment by sebastianbergmann** [comment]
No. Merging of code coverage data from separate runs is outside the scope of PHPUnit. The dedicated PHPCOV CLI tool exists for that.

**Comment by TheCelavi** [comment]
@sebastianbergmann Fair enough, I didn't know about that tool. Truth to be told, PHPUnit has a lot of tools under the hood which are not familiar to general population, no surprise why I missed that one. 

Now, I have managed to achieve my goal, I am getting combined coverage report, thanks for the pointers, tool is awesome!

One thing that I have noticed, which is really bad experience for me, is that last hook on which I can react is "AfterLastTestHook". That means that I am not informed when main coverage report is generated. 

Because of that, I have to execute two commands, `composer run phpunit` and `composer run generate-coverage-report`. 

If you reconsider to add a hook, per example, "AfterCodeCoverageReport", we could react on that hook and generate our custom coverage report on-the-fly without need to run additional command to generate coverage report.

That would mean that we have same experience within this edge case as well. 

It is not a deal breaker, but it would be nice to have. 

Of course, if you are willing to approve such feature, I would be more than happy to send PR, implementing "AfterCodeCoverageReport" hook would be quite trivial.

Once again, thank you for pointers.
